cs/src/core/Device/ManagedLocalStorageDevice.cs

```diff
@@ -97,65 +104,107 @@ private void RecoverFiles()
+                try
+                {
+                    Interlocked.Increment(ref numPending);
+                    streampool = GetOrAddHandle(segmentId).Item1;
+                    (logReadHandle, offset) = streampool.Get();
+                    logReadHandle.Seek((long)sourceAddress, SeekOrigin.Begin);
+#if NETSTANDARD2_1
+                    unsafe
                     {
-                        if (t.Exception.InnerException is IOException)
-                        {
-                            var e = t.Exception.InnerException as IOException;
-                            errorCode = (uint)(e.HResult & 0x0000FFFF);
-                        }
-                        else
-                        {
-                            errorCode = uint.MaxValue;
-                        }
+                        umm = new UnmanagedMemoryManager<byte>((byte*)destinationAddress, (int)readLength);
                     }
 
-                    // Sequentialize all reads from same handle on non-windows
-#if NETSTANDARD
-                    if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                    // FileStream.WriteAsync is not thread-safe hence need a lock here
+                    lock (this)
+                    {
+                        readTask = logReadHandle.ReadAsync(umm.Memory).AsTask();
+#else
+                    memory = pool.Get((int)readLength);
+                    lock (this)
+                    {
+                        readTask = logReadHandle.ReadAsync(memory.buffer, 0, (int)readLength);
+#endif
+                    }
+                    if (IsWindows)
                     {
-                        if (offset >= 0) streampool?.Return(offset);
+                        // If non-netstandard (==windows-only), or netstandard+windows, we can return to the pool immediately.
+                        // For netstandard+linux we will return after the operation completes.
+                        streampool?.Return(offset);
                     }
+                }
+                catch
+                {
+                    Interlocked.Decrement(ref numPending);
+#if !NETSTANDARD2_1
+                    memory?.Return();
 #endif
+                    streampool?.Return(offset);
+                    streampool?.DisposeIfNeeded(offset, logReadHandle);
+                    callback(uint.MaxValue, 0, context);
+                    return;
+                }
+                
+                try
+                {
+                    numBytes = await readTask;
 
-                    callback(errorCode, (uint)t.Result, context);
+                    // Sequentialize all reads from same handle on non-windows
+                    if (!IsWindows)
+                    {
+                        streampool?.Return(offset);
+                    }
+                    streampool?.DisposeIfNeeded(offset, logReadHandle);
                 }
-                );
+                catch (Exception ex)
```

Comment:
In case of any exceptions in await readTask, the streamPool.Return (when IsWindows=false) and streamPool.DisposeIfNeeded will not be invoked. Is that intentional?

Reply:
Fixed, please re-check code.